homeassistant/components/onvif/camera.py

```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
             _LOGGER.error("Couldn't setup camera '%s'. Error: %s", self._name, err)
 
-    def setup_ptz(self):
+    async def setup_ptz(self):
         """Set up PTZ if available."""
         _LOGGER.debug("Setting up the ONVIF PTZ service")
-        if self._camera.get_service("ptz", create=False) is None:
+        if self._camera.get_service("ptz") is None:
             _LOGGER.debug("PTZ is not available")
         else:
             self._ptz_service = self._camera.create_ptz_service()
-            _LOGGER.debug("Completed set up of the ONVIF camera component")
+        _LOGGER.debug("Completed set up of the ONVIF camera component")
 
-    async def async_perform_ptz(self, pan, tilt, zoom):
+    async def async_perform_ptz(
+        self, pan, tilt, zoom, distance, speed, move_mode, continuous_duration
+    ):
         """Perform a PTZ action on the camera."""
         if self._ptz_service is None:
             _LOGGER.warning("PTZ actions are not supported on camera '%s'", self._name)
             return
 
         if self._ptz_service:
-            pan_val = 1 if pan == DIR_RIGHT else -1 if pan == DIR_LEFT else 0
-            tilt_val = 1 if tilt == DIR_UP else -1 if tilt == DIR_DOWN else 0
-            zoom_val = 1 if zoom == ZOOM_IN else -1 if zoom == ZOOM_OUT else 0
-            req = {
-                "Velocity": {
-                    "PanTilt": {"_x": pan_val, "_y": tilt_val},
-                    "Zoom": {"_x": zoom_val},
-                }
-            }
+            pan_val = (
+                distance if pan == DIR_RIGHT else -distance if pan == DIR_LEFT else 0
+            )
+            tilt_val = (
+                distance if tilt == DIR_UP else -distance if tilt == DIR_DOWN else 0
+            )
+            zoom_val = (
+                distance if zoom == ZOOM_IN else -distance if zoom == ZOOM_OUT else 0
+            )
+            speed_val = speed
+            _LOGGER.debug(
+                "Calling %s PTZ | Pan = %4.2f | Tilt = %4.2f | Zoom = %4.2f | Speed = %4.2f",
+                move_mode,
+                pan_val,
+                tilt_val,
+                zoom_val,
+                speed_val,
+            )
             try:
-                _LOGGER.debug(
-                    "Calling PTZ | Pan = %d | Tilt = %d | Zoom = %d",
-                    pan_val,
-                    tilt_val,
-                    zoom_val,
-                )
-
-                await self._ptz_service.ContinuousMove(req)
+                req = self._ptz_service.create_type(move_mode)
+                req.ProfileToken = await self.async_obtain_profile_token()
+                if move_mode == CONTINUOUS_MOVE:
+                    req.Velocity = {
+                        "PanTilt": {"x": pan_val, "y": tilt_val},
+                        "Zoom": {"x": zoom_val},
+                    }
+
+                    await self._ptz_service.ContinuousMove(req)
+                    sleep(continuous_duration)
```

Comment:
Sync sleep isn't allowed in the event loop, inside a coroutine. Please await `asyncio.sleep` instead.

Reply:
Thanks !